### PR TITLE
Add numpy and hdf5 support to segmentation models

### PR DESCRIPTION
**Pull request by javier-alvarez**
- Add support for npy and npz arrays for segmentation models
- Add support for hdf5 files in segmentation models
- Change subjectIDs to strings

### CODE REVIEW

**Comment by philippwitte** [comment]
To use hdf5 with the `SegmentationModelBase`, there are a few checks in `full_image_dataset.py` that I had to disable for now:

- Line 263 throws an error because it finds multiple file extensions. This is because my image path has the form `path_to_file.h5:field_name:channel:label`.

- Line 244 expects nifti format

- Line 350 is checking whether the hdf5 file exists at the image path, which throws an error due to the above naming convention

- Line 285 expects nifti format

File: docs/creating_dataset.md [hunk 1]
```diff
@@ -19,7 +19,11 @@ InnerEye expects segmentation datasets to have the following structure:
  at minimum:
     * `subject`: A unique positive integer assigned to every patient
     * `channel`: The imaging channel or ground truth structure described by this row. 
-    * `filePath`: Path to the Nifti file for this scan or structure
+    * `filePath`: Path to the file for this scan or structure. We support nifti (nii.gz), numpy (npz) and hdf5(h5).
+        * For HDF5 path suffix with | separator
+            * For images <path>|<dataset_name>|<channel index>
+            * For segmentation binary <path>|<dataset_name>|<channel index>
+            * For segmentation multimap <path>|<dataset_name>|<channel index>|<multimap value>
```
Comment (ant0nsc) [inline docs/creating_dataset.md:26]: We also need to add here what assumption we make about the ordering of the array. Which index in the HDF5 array corresponds to X, Y, Z? Are we dropping any dimension? 
Segmentation binary and multimap should be better described.

File: docs/creating_dataset.md [hunk 2]
```diff
@@ -19,7 +19,11 @@ InnerEye expects segmentation datasets to have the following structure:
  at minimum:
     * `subject`: A unique positive integer assigned to every patient
     * `channel`: The imaging channel or ground truth structure described by this row. 
-    * `filePath`: Path to the Nifti file for this scan or structure
+    * `filePath`: Path to the file for this scan or structure. We support nifti (nii.gz), numpy (npz) and hdf5(h5).
+        * For HDF5 path suffix with | separator
+            * For images <path>|<dataset_name>|<channel index>
+            * For segmentation binary <path>|<dataset_name>|<channel index>
```
Comment (ant0nsc) [inline docs/creating_dataset.md:25]: what happens if what we think is a binary segmentaiton is actually not?

**Review by ant0nsc — CHANGES_REQUESTED** [review]
It's looking great overall, I've left comments where some polishing is still required.
I'd also suggest to improve test coverage for HDF5 loading. You are presently only testing the happy path (with the nifti-hdf5 conversion). If you make use of the mock libraries, you could also test the corner and failure cases (dataset_name does not exist, index out of range). Segmentation multimap loading is, I think, not tested either.

File: InnerEye/ML/utils/split_dataset.py [hunk 3]
```diff
@@ -215,16 +214,16 @@ def from_proportions(df: pd.DataFrame,
             proportion_test=proportion_test
         )
         return DatasetSplits.from_subject_ids(df,
-                                              list(ranges[ModelExecutionMode.TRAIN]),
-                                              list(ranges[ModelExecutionMode.TEST]),
-                                              list(ranges[ModelExecutionMode.VAL]),
+                                              list(map(str, ranges[ModelExecutionMode.TRAIN])),
```
[the remark this answers is not in the document]
Comment (javier-alvarez) [reply InnerEye/ML/utils/split_dataset.py:217]: correct

File: InnerEye/ML/reports/notebook_report.py [hunk 4]
```diff
@@ -48,7 +53,7 @@ def generate_notebook(template_notebook: Path, notebook_params: Dict, result_not
             resources=resources,
             notebook_name=str(result_notebook.stem)
         )
-    return result_notebook.with_suffix(resources['output_extension'])
+    return resources
```
Comment (ant0nsc) [inline InnerEye/ML/reports/notebook_report.py:56]: does nto need to return resources at all. Rather, it should return result_notebook.with_suffix(resources['output_extension'])

File: docs/creating_dataset.md (hunk 2 above)
Comment (javier-alvarez) [reply docs/creating_dataset.md:25]: returns all 0s

**Review by ant0nsc — APPROVED** [review]

**Review by Shruthi42 — APPROVED** [review]
